### PR TITLE
B2233 2234

### PR DESCRIPTION
**Pull request by ddumet**
## Description
Redmine 2233 & 2234

## How Has This Been Tested?
Thought experiment

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x ] New feature and improvements (non-breaking change which adds/improves functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Code Refactoring (internal change which is not user facing)
- [ ] Documentation
- [ ] Test, CI, or build

## Component(s) if applicable
- [ x] BentoService (service definition, dependency management, API input/output adapters)
- [ ] Model Artifact (model serialization, multi-framework support)
- [ ] Model Server (mico-batching, dockerisation, logging, OpenAPI, instruments)
- [ ] YataiService gRPC server (model registry, cloud deployment automation)
- [ ] YataiService web server (nodejs HTTP server and web UI)
- [ ] Internal (BentoML's own configuration, logging, utility, exception handling)
- [ ] BentoML CLI

## Checklist:

- [ ] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: bentoml/server/api_server.py [hunk 1]
```diff
@@ -454,6 +454,7 @@ def handle_ml_processing(self):
         logger.info('requests status_code for GET subscriptionQuery: %s', r.status_code)
         logger.info('Data: %s', r.json())
         ENTITY_INPUT_DATA = r.json()['entityID']['value']
+        ATTRIBUTE_INPUT_DATA = r.json()['query']['value'].split('=')[1]
```
Comment (bobeal) [inline bentoml/server/api_server.py]: I don't get this.

The watched attributes that will be used as inputs to the ML model are defined by the ML Model entity, not by the subscription query.

File: bentoml/server/api_server.py [hunk 2]
```diff
@@ -454,6 +454,7 @@ def handle_ml_processing(self):
         logger.info('requests status_code for GET subscriptionQuery: %s', r.status_code)
         logger.info('Data: %s', r.json())
         ENTITY_INPUT_DATA = r.json()['entityID']['value']
```
Comment (bobeal) [inline bentoml/server/api_server.py:457]: I did not really pay attention to this the first time, but it does not match the definition of a SubscriptionQuery entity.

File: bentoml/server/api_server.py [hunk 3]
```diff
@@ -468,7 +469,7 @@ def handle_ml_processing(self):
                     'type': 'River'
                 }
             ],
-            'watchedAttributes': ['precipitation'],
+            'watchedAttributes': [ATTRIBUTE_INPUT_DATA],
```
Comment (bobeal) [inline bentoml/server/api_server.py:472]: be careful, `precipitation` is still hardcoded in some other places in this file.

File: bentoml/server/api_server.py (hunk 1 above)
Comment (ddumet) [reply bentoml/server/api_server.py]: I'm confused. What is/should be in the subscriptionQuery entity if not the input data desired by the consumer app ? If the watched attributes are defined by the ML Model entity, what is the purpose of the subscriptionQuery ?

File: bentoml/server/api_server.py (hunk 2 above)
Comment (ddumet) [reply bentoml/server/api_server.py:457]: What is it then?

Comment (bobeal) [reply bentoml/server/api_server.py:457]: The answer is in the specification :)

File: bentoml/server/api_server.py (hunk 1 above)
Comment (bobeal) [reply bentoml/server/api_server.py]: The ML Model declares what it accepts as input data (e.g. a `precipitation` property), the subscription query says where this property will be found (e.g. in a specific `River` entity)

File: bentoml/server/api_server.py (hunk 3 above)
Comment (ddumet) [reply bentoml/server/api_server.py:472]: A different route. I don't know how to make it not hardcoded. Conf ? Querying MLModel if you stand by your first comment on the watched attributes being defined by the MLModel ?

File: bentoml/server/api_server.py (hunk 2 above)
Comment (ddumet) [reply bentoml/server/api_server.py:457]: What's the difference causes problem ?

{
  "id":"urn:ngsi-ld:SubscriptionQuery:01",
  "type":"SubscriptionQuery",
  "entities": [
    {
      "type": "WaterConsumption"
    }
  ],
  “q”: “refCity==urn:ngsi-ld:City:Valbonne” 
}


{
    'id': 'urn:ngsi-ld:MLProcessing:SubscriptionQuery:e7be459e-dcee-46ab-90da-fba3120db4ff',
    'type': 'https://uri.fiware.org/ns/data-models#SubscriptionQuery',
    'entityID': {
        'type': 'Property',
        'value': 'urn:ngsi-ld:River:014f5730-72ab-4554-a106-afbe5d4d9d26'
    },
    'query': {
        'type': 'Property',
        'value': '?attrs=precipitation'
    },
    '@context': ['https://uri.etsi.org/ngsi-ld/v1/ngsi-ld-core-context.jsonld']
}

File: bentoml/server/api_server.py (hunk 3 above)
Comment (bobeal) [reply bentoml/server/api_server.py:472]: Put it in conf, in the same way than the id of your ML model entity is configured. This is part of the definition of the ML model that is embedded in Bento, so this is not subject to change dynamically.

File: bentoml/configuration/default_bentoml.cfg [hunk 4]
```diff
@@ -99,5 +99,6 @@ bento_uri_default_expiration = 3000
 [ngsild]
 cb_url = https://stellio-dev.eglobalmark.com
 at_context = https://raw.githubusercontent.com/easy-global-market/ngsild-api-data-models/master/mlaas/jsonld-contexts/mlaas-precipitation-contexts.jsonld
-access_token = changeme
-ml_model_id=urn:ngsi-ld:MLModel:flow:predict
+access_token = MY_TOKEN_ID
+ml_model_id = urn:ngsi-ld:MLModel:flow:predict
+ml_model_input = precipitation
```
Comment (bobeal) [inline bentoml/configuration/default_bentoml.cfg:104]: I guess you could also have the `ml_model_output` here

Comment (ddumet) [reply bentoml/configuration/default_bentoml.cfg:104]: Possibly. Should I add it ?

File: bentoml/server/api_server.py (hunk 2 above)
Comment (bobeal) [reply bentoml/server/api_server.py:457]: The thing with the `SubscriptionQuery` in the specification is that it is the same syntax as a regular `Subscription`, so the user does not have to learn a new syntax to subscribe to something. But let's keep this for a later iteration.

File: bentoml/configuration/default_bentoml.cfg (hunk 4 above)
Comment (bobeal) [reply bentoml/configuration/default_bentoml.cfg:104]: Cleaner if both are extracted in conf.

File: bentoml/server/api_server.py (hunk 2 above)
Comment (ddumet) [reply bentoml/server/api_server.py:457]: I don't understand. It's better to talk.

**Pull request merged**